packages/nimble-components/src/rich-text/mention-list-box/index.ts

```diff
@@ -0,0 +1,294 @@
+     */
+    public clickHandler(e: MouseEvent): boolean {
+        // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
+        const captured = (e.target as HTMLElement).closest(
+            'option,[role=option]'
+        ) as ListOption | null;
+
+        if (!captured || captured.disabled) {
+            return false;
+        }
+        this.activateMention({
+            href: captured.value,
+            displayName: captured.innerText
+        } as MentionDetail);
+        return true;
+    }
+
+    public updateMentionExtensionConfig(mentionExtensionConfig?: MentionExtensionConfiguration[]): void {
+        this.mentionExtensionConfig = mentionExtensionConfig;
+        this.setActiveMappingConfigs();
+        this.filterOptions();
+    }
+
+    public onMention(props: SuggestionProps): void {
+        this.suggestionProps = props;
+        this.setActiveCharacter(props.text.slice(0, 1));
+        this.filter = props.query;
+        this.setAnchorElement(props.decorationNode as HTMLElement);
+        this.setOpen(true);
+        void this.selectFirstOptionIfValidOptionExists();
+    }
+
+    public close(): void {
+        this.setOpen(false);
+    }
+
+    /**
+     * @internal
+     */
+    public activeCharacterChanged(): void {
+        this.setActiveMappingConfigs();
+    }
+
+    /**
+     * @internal
+     */
+    public filterChanged(): void {
+        this.filterOptions();
+    }
+
+    /**
+     * @internal
+     */
+    public childItemsChanged(): void {
+        void this.selectFirstOptionIfValidOptionExists();
+    }
+
+    /**
+     * @internal
+     */
+    public anchorElementChanged(prev: HTMLElement, next: HTMLElement): void {2
+        if (prev) {
+            this.intersectionObserver.unobserve(prev);
+        }
+        if (this.region && this.anchorElement) {
+            this.region.anchorElement = this.anchorElement;
+            this.region.update();
+            this.intersectionObserver.observe(next);
+        }
+    }
+
+    private setActiveCharacter(activeCharacter: string): void {
+        this.activeCharacter = activeCharacter;
+    }
+
+    private setAnchorElement(anchorElement: HTMLElement): void {
+        this.anchorElement = anchorElement;
+    }
+
+    private scrollOptionIntoView(): void {
+        this.listBox?.firstSelectedOption?.scrollIntoView({ block: 'nearest' });
+    }
+
+    private setActiveMappingConfigs(): void {
+        this.activeMappingConfigs = this.activeCharacter ? this.mentionExtensionConfig?.find(config => config.character === this.activeCharacter)?.mappingConfigs : undefined;
+    }
+
+    private filterOptions(): void {
+        if (!this.childItems || this.filter === undefined) {
+            return;
+        }
+        const normalizedFilter = normalizeString(this.filter);
+        this.childItems.forEach(element => {
+            const listOption = element as ListOption;
+            const normalizedText = normalizeString(listOption.text);
```

Comment:
This seems like it could be slow. Every time the user types another character, the filter changes and we call `filterOptions`, which normalizes every list option. If there are dozens or hundreds of users, it might cause a noticeable delay. We should try this with lots of users in the list and see how it performs. If it's too slow, maybe we cache the list of normalized values, or we reconsider whether it is necessary to do a normalized match.

Reply:
As we discussed, the clients are expected to filter and supply the mapping elements. Hence, this will not be a concern within the nimble layer.

Reply:
Are we expecting the client to change the mapping elements on every change in the filter? If so, does that mean we don't need to filter in the editor at all?

Reply:
Client is expected to reduce the set of mappings to all the mappings already used + the mappings in the dropdown. So the expectations are pretty complex.

Slot only a few mentioned mappings (current mention mappings + a few more up to 10 ish)
On mention-update use the updated value to slot all currently mentioned mappings + whatever is in the list

So the filter needs to run so that the mappings that already mentioned don't show up in the list.

Reply:
To capture the performance of the `diacriticInsensitiveStringNormalizer` method, I have created a code snippet that attempts to normalize 1000 user names.

The operation was completed within 0.5 to 1.5 ms on average.

Attaching the code snippet and the gif for reference.

```
let diacriticInsensitiveStringNormalizer = function(value) {
    return value
        .normalize('NFD')
        .replace(/[\u0300-\u036f]/g, '')
        .toLowerCase();
};

// holds 50 sample users
let sampleUsers = [
  "Alice Bérnard", "Böb Müller", "Chàrlie Dupont", "Dávid Sánchez", "Ëmma Jøhnsøn", "Fränk Dávïs", "Grâce Míller", 
  "Hénry Wilsön", "Ïvy Lée", "Jâck Thompšon", "Kátie Táylor", "Léo Märtínez", "Nöah Gómez", "Olívia Léblanc", 
  "Pénélope Dubois", "Quèntin Roüx", "Róse Mórgan", "Sàrah López", "Théodore Nguyên", "Ürsula Wagner", 
  "Víctor Íbrahim", "Wén Huáng", "Xàvier Delacroix", "Yüko Yamamoto", "Zoé Carré", "Andréa Fischer", 
  "Bèatrice Dupuis", "Cédric Lefèvre", "Délphine Martin", "Étienne Moreau", "Flörence Lambert", "Gabriel Lévesque", 
  "Hélène Dubois", "Isaïah André", "Jülia Costa", "Kévin Lemoine", "Lïna Deschamps", "Màrc Dupont", "Nàdia Rocha", 
  "Oscar Léger", "Päuline Girard", "Quincy Adams", "Rëbecca Märcel", "Sébastien Gagnon", "Tátiana Ruiz", "Ünique Léon",
  "Véra Leblanc", "William Martin", "Xénia Rodriguez", "Yánnis Papadopoulos"
]

let inputUsers = [];
let outputUsers = [];

// Generates 1000 users
for (let i = 0; i < 20; i++) {
  inputUsers = [...inputUsers, ...sampleUsers];
}

console.log("Sample count", inputUsers.length);

console.time("Measure");
for(let user of inputUsers) {
    outputUsers.push(diacriticInsensitiveStringNormalizer(user));
}
console.timeEnd("Measure");

console.log(outputUsers);

```


![Normalization Metrics](https://github.com/ni/nimble/assets/123383804/67173434-b013-480f-8608-3c193aa0b109)

Reply:
Updated the usage docs on restricting the number of users at a time to improve the performance.